pxr/imaging/hdSt/testenv/testHdStMaterialXShaderGen.cpp

```diff
@@ -28,16 +28,73 @@
 
 #include "pxr/base/tf/diagnostic.h"
 
+#include <MaterialXGenShader/Util.h>
+
 #include <iostream>
 
 PXR_NAMESPACE_USING_DIRECTIVE
 
 namespace mx = MaterialX;
 
+bool IsDifferentFrom(mx::InputPtr const& input, float value) {
+    if (!input) {
+        return false;
+    }
+
+    if (input->hasValue()) {
+        auto val = input->getValue();
+        if (val->isA<float>()) {
+            return val->asA<float>() != value;
+        }
+        return true;
+    }
+
+    return input->hasNodeName() || input->hasNodeGraphString() ||
+           input->hasOutputString() || input->hasInterfaceName();
+}
+
+bool IsTranslucentDocument(mx::DocumentPtr const& mxDoc)
+{
+    // Find renderable elements in the Mtlx Document.
+    std::vector<mx::TypedElementPtr> renderableElements;
+    mx::findRenderableElements(mxDoc, renderableElements);
+
+    // Should have exactly one renderable element (material).
+    if (renderableElements.size() != 1) {
+        TF_CODING_ERROR("Generated MaterialX Document does not "
+                        "have 1 material");
+        return false;
+    }
+
+    // Extract out the Surface Shader Node for the Material Node
+    mx::TypedElementPtr renderableElem = renderableElements.at(0);
+    mx::NodePtr node = renderableElem->asA<mx::Node>();
+    if (node && node->getType() == mx::MATERIAL_TYPE_STRING) {
+        // Use auto so can compile against MaterialX 1.38.0 or 1.38.1
+        auto mxShaderNodes = mx::getShaderNodes(node, mx::SURFACE_SHADER_TYPE_STRING);
+        if (!mxShaderNodes.empty()) {
+            renderableElem = *mxShaderNodes.begin();
+        }
+    }
+
+    auto renderableNode = renderableElem->asA<mx::Node>();
+    if (renderableNode && renderableNode->getCategory() == "UsdPreviewSurface")
```

Comment:
Sadly, `mx::isTransparentSurface` is only able to handle the most basic cases. It digs inside layers of NodeGraph until it finds either the opacity input of the ND_surface, or the weight input of a transparent Schlick node.

Then is has to traverse back up the connections to the calling node interface to check what was the actual value.

The problem is when there are intermediate nodes in that code path. MaterialX is able to deal with the luminance found in standard_surface that converts the color3 interface opacity input down to the float opacity input of the ND_surface node, but when the graph is more complex, as is the case with PreviewSurface and glTf, it becomes very hard to evaluate the expression in C++ to know or not if the surface is transparent.

In these cases, the `mx::isTransparentSurface` will always return transparent, and this is why we need a custom test. This is also why I had to fix the output of a few PreviewSurface unit tests because the surface was not transparent anymore.